sdk/azcore/arm/resource_identifier.go

```diff
@@ -0,0 +1,210 @@
+//go:build go1.16
+// +build go1.16
+
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package arm
+
+import (
+	"fmt"
+	"strings"
+)
+
+const (
+	providersKey             = "providers"
+	subscriptionsKey         = "subscriptions"
+	resourceGroupsLowerKey   = "resourcegroups"
+	locationsKey             = "locations"
+	builtInResourceNamespace = "Microsoft.Resources"
+)
+
+var (
+	// RootResourceIdentifier defines a ResourceID of a tenant as a root level parent of all other ResourceID
+	RootResourceIdentifier = &ResourceID{
+		Parent:       nil,
+		ResourceType: TenantResourceType,
+		Name:         "",
+	}
+)
+
+// ResourceID represents a resource ID such as `/subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/myRg`
+// Don't create this type directly, use ParseResourceID() instead
+type ResourceID struct {
+	Parent            *ResourceID
+	SubscriptionId    string
+	Provider          string
+	ResourceGroupName string
+	Location          string
+	ResourceType      ResourceType
+	Name              string
+
+	isChild     bool
+	stringValue string
+}
+
+func newResourceID(parent *ResourceID, resourceTypeName string, resourceName string) *ResourceID {
```

Comment:
A minor thing, I try to keep the unexported code after all the exported code to make the public stuff easier to find.

Reply:
I will move these unexported helpers at the end of this file